joeyHF: adapting joeyNMT for Masakhane Benchmark Models on Spaces

chrisemezue opened the pull request:
My main work was in the `prediction.py` file. 

I created a `load_params_for_prediction` function which loads the parameters and model needed to perform translation and outputs the model parameters -> `params`.

I then tweaked the `translate,py` file to take in `params` and make translation.

Any idea you have for me to do to not make it conflict with the joeyNMT's `translate,py` file will be very much appreciated.

juliakreutzer commented:
Hi @chrisemezue!
Thanks for the code! I think it needs some refactoring before we can merge it, in order to make it easy to read and use and compatible with the rest of JoeyNMT. I'll add detailed comments to the code.
In the meantime, can you please try running
`pytest` and `pylint --rcfile=.pylintrc joeynmt`. The output of pylint should give you some hints how to make the code cleaner, and pytest will flag any unit test issues. Both need to pass without errors before we can merge.
Eventually, your addition should also be tested through a unit test, please add it to `https://github.com/joeynmt/joeynmt/blob/main/test/unit/test_prediction.py`.
Thanks!

juliakreutzer commented on joeynmt/prediction.py:
```diff
@@ -268,19 +268,25 @@ def test(cfg_file,
     translations and storing them and attention plots.
 
-    :param cfg_file: path to configuration file
+    :param cfg_file: path to configuration file. Config file can be the yaml file or the loaded config file
```
This ambiguity would be helpful to express in the [typing information](https://docs.python.org/3/library/typing.html#) above in line 261.

juliakreutzer commented on joeynmt/prediction.py:
```diff
@@ -268,19 +268,25 @@ def test(cfg_file,
+        cfg=cfg_file
     model_dir = cfg["training"]["model_dir"]
 
+    '''
```
Can you please add a comment here why this block is needed?

juliakreutzer commented on joeynmt/prediction.py:
```diff
@@ -382,76 +388,39 @@ def test(cfg_file,
+def load_params_for_prediction(cfg_file,ckpt: str):
+
     """
     Interactive translation function.
```
Remove this line, seems outdated.

juliakreutzer commented on joeynmt/prediction.py:
```diff
@@ -382,76 +388,39 @@ def test(cfg_file,
-    :param output_path: path to output file
-    :param batch_class: class type of batch
-    :param n_best: amount of candidates to display
+    :param cfg_file: path to configuration file/ configuration file dictionary to load parameters
```
Typing, see above.

juliakreutzer commented on joeynmt/prediction.py:
```diff
@@ -382,76 +388,39 @@ def test(cfg_file,
-        ckpt = get_latest_checkpoint(model_dir)
+    
+    params_dict={}
+    if isinstance(cfg_file,str):
```
This block is duplicated - maybe it's worth factoring it out into a separate function in `helpers.py`.

juliakreutzer commented on joeynmt/prediction.py:
```diff
@@ -382,76 +388,39 @@ def test(cfg_file,
+   
 
     # read vocabs
-    src_vocab_file = cfg["data"].get("src_vocab", model_dir + "/src_vocab.txt")
```
Why this reduction of flexibility?

juliakreutzer commented on joeynmt/prediction.py:
```diff
@@ -382,76 +388,39 @@ def test(cfg_file,
-              output_path: str = None,
-              batch_class: Batch = Batch,
-              n_best: int = 1) -> None:
+def load_params_for_prediction(cfg_file,ckpt: str):
```
Please add documentation + typing for what this function returns. Thanks!

juliakreutzer replied:
It's important to clarify that the function loads the hyperparameters from the config AND loads the model (parameters).

juliakreutzer commented on joeynmt/prediction.py:
```diff
@@ -460,13 +429,40 @@ def _translate_data(test_data):
     # parse test args
     batch_size, batch_type, use_cuda, device, n_gpu, level, _, \
         max_output_length, beam_size, beam_alpha, postprocess, \
         bpe_type, sacrebleu, _, _ = parse_test_args(cfg, mode="translate")
```
How about returning the dictionary directly by the `parse_test_args` function? This would save us the repeated packing/unpacking.

juliakreutzer commented on joeynmt/prediction.py:
```diff
@@ -460,13 +429,40 @@ def _translate_data(test_data):
+    if ckpt is None:
+        raise Exception('Model checkpoint must be given')
     # load model state from disk
-    logger.info("Loading model from %s", ckpt)
```
Why is this removed?

juliakreutzer commented on joeynmt/prediction.py:
```diff
@@ -475,10 +471,52 @@ def _translate_data(test_data):
+
+    
+    
+def translate(params,data,type_:int,output_path=None) -> None:
```
`type_` as a variable name and as int type is not very expressive: it's not clear from the documentation of the function what it means. 
Please add docstrings to the function.

juliakreutzer commented on joeynmt/prediction.py:
```diff
@@ -504,30 +542,28 @@ def write_to_file(output_path_set, hypotheses):
+        # single sentence is given
         batch_size = 1
         batch_type = "sentence"
-        while True:
```
What happened to this interactive mode now? We'd like to keep this as a feature. Perhaps this requires breaking down the translate function into two separate functions.

chrisemezue commented:
@juliakreutzer I have made final edits to the code based on your reviews. 
- I took all your reviews into consideration and worked on them.
- I ran the `pytest` and they worked with success. I added the test cases in `https://github.com/joeynmt/joeynmt/blob/main/test/unit/test_prediction.py`. 
- `pylint` ran and I corrected based on its reviews.

juliakreutzer commented:
Hi @chrisemezue, thank you so much for the changes! For the remaining requested changes (unresolved comments), did you address them as well or shall we discuss them more?
I have discussed with @may- and we will first merge a solution that doesn't require torchtext, and then revisit your PR. In that way we can make sure that everything works as expected and the dependency issues are resolved.

chrisemezue commented:
> Hi @chrisemezue, thank you so much for the changes! For the remaining requested changes (unresolved comments), did you address them as well or shall we discuss them more? I have discussed with @may- and we will first merge a solution that doesn't require torchtext, and then revisit your PR. In that way we can make sure that everything works as expected and the dependency issues are resolved.

I worked on all your comments @juliakreutzer .
Please let me know if there is anywhere I can help.

juliakreutzer commented:
Hi @chrisemezue, can you try a merge, now that Joey v2.0 is out?

chrisemezue commented:
Okay @juliakreutzer 
I see some conflicts. I will add it to my todo - to resolve the conflicts from my end.

chrisemezue commented:
> Hi @chrisemezue, can you try a merge, now that Joey v2.0 is out?

Hi @juliakreutzer . Please I am quite confused what I am to do now in order to merge it.

juliakreutzer commented:
Hi @chrisemezue can you try to resolve these? https://github.com/joeynmt/joeynmt/pull/164/conflicts
And test if it runs?

NimaBoscarino commented:
Hey there! Are there still conflicts in this PR? When I visit https://github.com/joeynmt/joeynmt/pull/164/conflicts I get a 404. I'm also over at Hugging Face with @chrisemezue, so I'm happy to help out to get this merged!

may- commented:
See #213
